my_2d_robot/src/joint_states_subscriber.cpp

```diff
@@ -1,19 +1,150 @@
+  // String used to store package filepath.
+  std::string my_output_;
+  // String used to store parsed data from yaml-cpp library.
+  std::string output_data_ = "";
+
+  // Callback method to get joint angles.
+  void joint_statesCallback(const sensor_msgs::JointState::ConstPtr& msg);
+
+  // Method to write the joint state angles and end effector position to yaml file.
+  void saveYaml();
+
+  // Method to calculate the position of the end effector.
+  Vector2f eePos(float joint_1, float joint_2)
+  {
+    // Includes the offsets for joint angles.
+    float Theta1 = joint_1 - theta1_offset_;
+    float Theta2 = joint_2 - theta2_offset_;
+
+    // Calculates the end effector position.
+    Vector2f position;
+    position << (link_1_ * cos(Theta1 * M_PI / 180)) + (link_2_ * cos((Theta1 + Theta2) * M_PI / 180)),
+        (link_1_ * sin(Theta1 * M_PI / 180)) + (link_2_ * sin((Theta1 + Theta2) * M_PI / 180));
+
+    return position;
+  }
+};
+
+// Class methods.
+// Method to get the joint state information from publisher node.
+void SensorMeasurementData::joint_statesCallback(const sensor_msgs::JointState::ConstPtr& msg)
 {
-  ROS_INFO("I heard: [%s]", msg->data.c_str());
+  ROS_INFO("I heard: [%F]", msg->position[0]);
+
+  position_joint1_ = msg->position[0];
+  position_joint2_ = msg->position[1];
+
+  ee_position_ = eePos(position_joint1_, position_joint2_);
+
+  saveYaml();
 }
 
-int main (int argc, char **argv){
+// Method to write the joint state angles and end effector position to yaml file.
+void SensorMeasurementData::saveYaml()
+{
+  YAML::Emitter d_output;
+  d_output << YAML::BeginMap;
+  d_output << YAML::Key << "joint1 angle";
+  d_output << YAML::Value << position_joint1_;
+  d_output << YAML::Key << "joint2 angle";
+  d_output << YAML::Value << position_joint2_;
+  d_output << YAML::Key << "end effector x position";
+  d_output << YAML::Value << ee_position_(0);
+  d_output << YAML::Key << "end effector y position";
+  d_output << YAML::Value << ee_position_(1);
+  d_output << YAML::EndMap;
+
+  // Gets the package filepath.
+  my_output_ = ros::package::getPath("my_2d_robot");
 
-    ros::init(argc, argv, "joint_states_subscriber");
-    ros::NodeHandle n;
-    ros::Subscriber sub = n.subscribe("joint_states", 1000, joint_statesCallback);
+  // Checks if Output_yaml folder exists in package.
+  std::string check = my_output_.append("/Output_yaml");
+
+  // Creates folder if it does not already exist.
+  if (!(boost::filesystem::exists(check)))
+  {
+    boost::filesystem::create_directory(check);
+  }
+
+  // Amends filepath for storing yaml file.
+  my_output_ = check;
+  my_output_.append("/my_output.yaml");
+
+  output_data_.append(d_output.c_str());
+  output_data_.append("\n");
+}
+
+// Main function.
+int main(int argc, char** argv)
+{
+  ros::init(argc, argv, "joint_states_subscriber");
+  ros::NodeHandle n;
 
-    ros::spin();
+  SensorMeasurementData my_2d_robo_states = SensorMeasurementData(&n);
 
+  ros::spin();
```

Comment:
Correct me if I'm wrong. But it looks like your subscriber spins forever and only terminates on a Ctrl+C.

We should be able to specify a certain number of data points that we want in the launch file, your subscriber reads this in from the rosparam, and maintains an internal count in a class variable. When count == num_data points then the subscriber should immediately terminate. This will make data gen easier and is good design.